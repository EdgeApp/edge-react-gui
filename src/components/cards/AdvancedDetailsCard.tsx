import { EdgeTransaction } from 'edge-core-js'
import React, { PureComponent } from 'react'

import { lstrings } from '../../locales/strings'
import { openBrowserUri } from '../../util/WebUtils'
import { EdgeRow } from '../rows/EdgeRow'
import { ThemeProps, withTheme } from '../services/ThemeContext'
import { EdgeCard } from './EdgeCard'

const localizedFeeText = {
  satPerVByte: lstrings.transaction_details_advance_details_satpervbyte,
  gasPrice: lstrings.transaction_details_advance_details_gasprice,
  minerTip: lstrings.transaction_details_advance_details_minertip,
  gasLimit: lstrings.transaction_details_advance_details_gaslimit
}

const feeString = {
  high: lstrings.mining_fee_high_label_choice,
  standard: lstrings.mining_fee_standard_label_choice,
  low: lstrings.mining_fee_low_label_choice
}

interface OwnProps {
  transaction: EdgeTransaction
  url?: string
}

type Props = OwnProps & ThemeProps

export class AdvancedDetailsCardComponent extends PureComponent<Props> {
  getRecipientAddress = () => (this.props.transaction.spendTargets ? this.props.transaction.spendTargets[0].publicAddress : '')

  openUrl = () => {
    const { url } = this.props
    if (url == null || url === '') return
    openBrowserUri(url)
  }

  openProveUrl = () => {
    const { txid, txSecret } = this.props.transaction
    const recipientAddress = this.getRecipientAddress()
    if (recipientAddress === '' || txid === '' || txSecret == null) return
    const url = `https://blockchair.com/monero/transaction/${txid}?address=${recipientAddress}&viewkey=${txSecret}&txprove=1`
    openBrowserUri(url)
  }

  renderFeeOptions(): string {
    const { networkFeeOption, requestedCustomFee } = this.props.transaction

    if (networkFeeOption === 'custom') {
      return `${lstrings.mining_fee_custom_label_choice}\n${this.renderFees(requestedCustomFee)}`
    }

    return networkFeeOption != null ? feeString[networkFeeOption] : lstrings.mining_fee_standard_label_choice
  }

  renderFees(fees: any = {}): string {
    let feeValueText = ''

    for (const feeKey of Object.keys(fees)) {
      // @ts-expect-error
      const feeFullString = `${localizedFeeText[feeKey] ?? feeKey} ${fees[feeKey]}`
      feeValueText = feeValueText === '' ? feeValueText + feeFullString : feeValueText + `\n${feeFullString}`
    }

    return feeValueText
  }

  render() {
    const { url } = this.props
    const { feeRateUsed, networkFeeOption, ourReceiveAddresses, signedTx, txid, txSecret, deviceDescription } = this.props.transaction
    const recipientAddress = this.getRecipientAddress()
    let receiveAddressesString
    if (ourReceiveAddresses != null && ourReceiveAddresses.length > 0) {
      receiveAddressesString = ourReceiveAddresses.join('\n')
    }

    return (
      <EdgeCard sections>
        {url != null && url !== '' && (
          <EdgeRow
            rightButtonType="touchable"
            title={lstrings.transaction_details_view_advanced_data}
            body={lstrings.transaction_details_advance_details_show_explorer}
            onPress={this.openUrl}
          />
        )}
        {receiveAddressesString != null && <EdgeRow rightButtonType="copy" title={lstrings.my_receive_addresses_title} body={receiveAddressesString} />}
        {networkFeeOption != null && <EdgeRow title={lstrings.transaction_details_advance_details_fee_setting} body={this.renderFeeOptions()} />}
        {feeRateUsed != null && <EdgeRow title={lstrings.transaction_details_advance_details_fee_used} body={this.renderFees(feeRateUsed)} />}
        {txSecret != null && <EdgeRow rightButtonType="copy" title={lstrings.transaction_details_advance_details_txSecret} body={txSecret} />}
        {txSecret != null && recipientAddress !== '' && txid !== '' && (
          <EdgeRow
            rightButtonType="touchable"
            title={lstrings.transaction_details_advance_details_payment_proof}
            body={lstrings.transaction_details_advance_details_show_explorer}
            onPress={this.openProveUrl}
          />
        )}
        {signedTx != null && signedTx !== '' ? (
          <EdgeRow rightButtonType="copy" title={lstrings.transaction_details_advance_details_raw_txbytes} body={signedTx} maximumHeight="small" />
        ) : null}

        {deviceDescription != null && <EdgeRow title={lstrings.transaction_details_advance_details_device} body={deviceDescription} />}
      </EdgeCard>
    )
  }
}

export const AdvancedDetailsCard = withTheme(AdvancedDetailsCardComponent)
