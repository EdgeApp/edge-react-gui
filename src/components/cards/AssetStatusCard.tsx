import { AssetStatus } from 'edge-info-server/types'
import * as React from 'react'
import { Platform } from 'react-native'
import IonIcon from 'react-native-vector-icons/Ionicons'

import { getLocaleOrDefaultString } from '../../locales/intl'
import { openBrowserUri } from '../../util/WebUtils'
import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'
import { IconMessageCard } from './IconMessageCard'
export const AssetStatusCard = (props: { assetStatus: AssetStatus }) => {
  const { statusType, localeStatusTitle, localeStatusBody, iconUrl, statusUrl, statusStartIsoDate, statusEndIsoDate } = props.assetStatus
  const theme = useTheme()
  const styles = getStyles(theme)

  const curDate = new Date().toISOString()
  const isWithinDate = statusStartIsoDate != null && statusEndIsoDate != null && statusStartIsoDate <= curDate && statusEndIsoDate >= curDate

  const title = getLocaleOrDefaultString(localeStatusTitle)
  const message = getLocaleOrDefaultString(localeStatusBody)
  const isValidText = title != null && message != null

  return isWithinDate && isValidText ? (
    <IconMessageCard
      message={message}
      title={title}
      testIds={{ title: 'statusCardTitle', message: 'statusCardMessage', close: 'statusCardClose' }}
      iconOrUri={
        // If not explicitly set, auto-fill if warning status.
        iconUrl == null ? (
          statusType === 'warning' ? (
            <IonIcon
              name={Platform.OS === 'ios' ? 'ios-warning-outline' : 'md-warning-outline'}
              style={styles.icon}
              color={theme.warningText}
              size={theme.rem(3)}
            />
          ) : (
            // statusType === 'info'
            <IonIcon name="information-circle-outline" size={theme.rem(1.25)} color={theme.warningText} />
          )
        ) : (
          iconUrl
        )
      }
      onPress={statusUrl ? () => openBrowserUri(statusUrl) : undefined}
    />
  ) : null
}

const getStyles = cacheStyles((theme: Theme) => ({
  icon: {
    width: theme.rem(3),
    height: theme.rem(3),
    marginRight: theme.rem(0.5)
  }
}))
