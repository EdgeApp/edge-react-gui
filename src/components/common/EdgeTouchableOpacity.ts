import { TouchableOpacity } from 'react-native'

import { withExtendedTouchable } from '../hoc/withExtendedTouchable'

export const EdgeTouchableOpacity = withExtendedTouchable(TouchableOpacity)
