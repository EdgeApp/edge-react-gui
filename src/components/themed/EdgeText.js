// @flow

import * as React from 'react'
import { Platform, StyleSheet, Text } from 'react-native'

import { type Theme, type ThemeProps, cacheStyles, withTheme } from '../services/ThemeContext.js'

type OwnProps = {|
  children: React.Node,
  ellipsizeMode?: string,
  numberOfLines?: number,
  style?: StyleSheet.Styles,
  disableFontScaling?: boolean,
  minimumFontScale?: number
|}

export class EdgeTextComponent extends React.PureComponent<OwnProps & ThemeProps & Text.props> {
  render() {
    const { children, style, theme, disableFontScaling = false, ...props } = this.props
    const { text, androidAdjust } = getStyles(theme)
    let numberOfLines = this.props.numberOfLines || 1
    if (typeof children === 'string' && children.includes('\n')) {
      numberOfLines = numberOfLines + (children.match(/\n/g) || []).length
    }

    return (
      <Text
        style={[text, style, Platform.OS === 'android' ? androidAdjust : null]}
        numberOfLines={numberOfLines}
        adjustsFontSizeToFit={!disableFontScaling}
        minimumFontScale={0.65}
        {...props}
      >
        {children}
      </Text>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  text: {
    color: theme.primaryText,
    fontFamily: theme.fontFaceDefault,
    fontSize: theme.rem(1),
    includeFontPadding: false
  },
  androidAdjust: {
    top: -1
  }
}))

export const EdgeText = withTheme(EdgeTextComponent)
