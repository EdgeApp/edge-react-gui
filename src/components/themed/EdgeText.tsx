import * as React from 'react'
import { Platform, Text, TextProps, TextStyle } from 'react-native'

import { cacheStyles, Theme, ThemeProps, withTheme } from '../services/ThemeContext'

interface OwnProps {
  children: React.ReactNode
  ellipsizeMode?: 'head' | 'middle' | 'tail' | 'clip'
  numberOfLines?: number
  style?: TextStyle
  disableFontScaling?: boolean
  minimumFontScale?: number
}

export class EdgeTextComponent extends React.PureComponent<OwnProps & ThemeProps & TextProps> {
  render() {
    const { children, style, theme, disableFontScaling = false, ...props } = this.props
    const { text, androidAdjust } = getStyles(theme)
    let numberOfLines = this.props.numberOfLines || 1
    if (typeof children === 'string' && children.includes('\n')) {
      numberOfLines = numberOfLines + (children.match(/\n/g) || []).length
    }

    return (
      <Text
        style={[text, style, Platform.OS === 'android' ? androidAdjust : null]}
        numberOfLines={numberOfLines}
        adjustsFontSizeToFit={!disableFontScaling}
        minimumFontScale={0.65}
        {...props}
      >
        {children}
      </Text>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  text: {
    color: theme.primaryText,
    fontFamily: theme.fontFaceDefault,
    fontSize: theme.rem(1),
    includeFontPadding: false
  },
  androidAdjust: {
    top: -1
  }
}))

export const EdgeText = withTheme(EdgeTextComponent)
