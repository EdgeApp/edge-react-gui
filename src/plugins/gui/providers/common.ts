import { asObject, asString } from 'cleaners'

import { FiatPluginRegionCode } from '../fiatPluginTypes'
import { FiatProviderError, FiatProviderSupportedRegions } from '../fiatProviderTypes'

export const RETURN_URL_SUCCESS = 'https://edge.app/redirect/success/'
export const RETURN_URL_FAIL = 'https://edge.app/redirect/fail/'
export const RETURN_URL_CANCEL = 'https://edge.app/redirect/cancel/'
export const RETURN_URL_PAYMENT = 'https://edge.app/redirect/payment/'

export const NOT_SUCCESS_TOAST_HIDE_MS = 5000

/**
 * @param providerId Unique providerId to filter.
 * @param regionCode FiatPluginRegionCode object with country & state/province
 * @param supportedRegions FiatProviderSupportedRegions object
 * @returns void if supported, throws otherwise
 */
export const validateRegion = (providerId: string, regionCode: FiatPluginRegionCode, supportedRegions: FiatProviderSupportedRegions): void => {
  const { countryCode, stateProvinceCode } = regionCode
  const countryInfo = supportedRegions[countryCode]
  if (countryInfo == null) return

  const { forStateProvinces, notStateProvinces } = countryInfo

  if (stateProvinceCode != null) {
    if (notStateProvinces?.includes(stateProvinceCode) === true) {
      throw new FiatProviderError({ providerId, errorType: 'regionRestricted' })
    }
    if (forStateProvinces?.includes(stateProvinceCode) === false) {
      throw new FiatProviderError({ providerId, errorType: 'regionRestricted' })
    }
  }
}

export const asStandardApiKeys = asObject({
  apiKey: asString
})
