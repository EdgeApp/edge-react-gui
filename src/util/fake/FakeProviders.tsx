import * as React from 'react'
import { Metrics, SafeAreaProvider } from 'react-native-safe-area-context'
import { Provider } from 'react-redux'
import { applyMiddleware, createStore } from 'redux'
import thunk from 'redux-thunk'

import { rootReducer, RootState } from '../../reducers/RootReducer'

type DeepPartial<T> = T extends object
  ? {
      [P in keyof T]?: DeepPartial<T[P]>
    }
  : T

export type FakeState = DeepPartial<RootState>

interface Props {
  children: React.ReactNode
  initialState?: FakeState
}

export function FakeProviders(props: Props) {
  const { children, initialState = {} } = props

  const store = React.useMemo(() => createStore(rootReducer, initialState as any, applyMiddleware(thunk)), [initialState])
  return (
    <SafeAreaProvider initialMetrics={initialMetrics}>
      <Provider store={store}>{children}</Provider>
    </SafeAreaProvider>
  )
}

// These match what our old react-native-safe-area-view library returned:
const initialMetrics: Metrics = {
  frame: { height: 1334, width: 750, x: 0, y: 0 },
  insets: { bottom: 0, left: 0, right: 0, top: 20 }
}
