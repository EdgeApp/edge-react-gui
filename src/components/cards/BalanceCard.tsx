import { lt } from 'biggystring'
import * as React from 'react'
import { LayoutChangeEvent, Text, View } from 'react-native'
import IonIcon from 'react-native-vector-icons/Ionicons'

import { toggleAccountBalanceVisibility } from '../../actions/LocalSettingsActions'
import { getFiatSymbol } from '../../constants/WalletAndCurrencyConstants'
import { useHandler } from '../../hooks/useHandler'
import { useWatch } from '../../hooks/useWatch'
import { formatNumber } from '../../locales/intl'
import { lstrings } from '../../locales/strings'
import { useDispatch, useSelector } from '../../types/reactRedux'
import { NavigationBase } from '../../types/routerTypes'
import { getTotalFiatAmountFromExchangeRates, removeIsoPrefix, zeroString } from '../../util/utils'
import { ButtonsView } from '../buttons/ButtonsView'
import { AnimatedNumber } from '../common/AnimatedNumber'
import { EdgeTouchableOpacity } from '../common/EdgeTouchableOpacity'
import { TransferModal } from '../modals/TransferModal'
import { Airship } from '../services/AirshipInstance'
import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'
import { EdgeText } from '../themed/EdgeText'
import { EdgeCard } from './EdgeCard'

// Numbers larger than this are likely to overflow the display width so don't
// use regular Text components which can auto shrink
const MAX_ANIMATED_AMOUNT = '10000000'
interface Props {
  navigation: NavigationBase
  onViewAssetsPress?: () => void
}

/**
 * Card that displays balance, deposit/send buttons, and a link to view assets
 */
export const BalanceCardUi4 = (props: Props) => {
  const { navigation, onViewAssetsPress } = props

  const dispatch = useDispatch()
  const theme = useTheme()
  const styles = getStyles(theme)

  const account = useSelector(state => state.core.account)
  const isBalanceVisible = useSelector(state => state.ui.settings.isAccountBalanceVisible)
  const defaultIsoFiat = useSelector(state => state.ui.settings.defaultIsoFiat)
  const fiatAmount = useSelector(state => getTotalFiatAmountFromExchangeRates(state, defaultIsoFiat))
  const exchangeRates = useSelector(state => state.exchangeRates)

  const activeWalletIds = useWatch(account, 'activeWalletIds')
  const currencyWallets = useWatch(account, 'currencyWallets')
  const currencyWalletErrors = useWatch(account, 'currencyWalletErrors')

  const exchangeRatesReady = React.useMemo(
    () =>
      activeWalletIds.every(walletId => {
        // Ignore wallets that have crashed:
        if (currencyWalletErrors[walletId]) return true

        // Both the wallet and its rate need to be loaded:
        const wallet = currencyWallets[walletId]
        return wallet != null && exchangeRates[`${wallet.currencyInfo.currencyCode}_${defaultIsoFiat}`] != null
      }),
    [activeWalletIds, currencyWalletErrors, currencyWallets, exchangeRates, defaultIsoFiat]
  )
  const [digitHeight, setDigitHeight] = React.useState(0)

  const fiatSymbol = defaultIsoFiat ? getFiatSymbol(defaultIsoFiat) : ''
  const fiatCurrencyCode = removeIsoPrefix(defaultIsoFiat)
  const formattedFiat = isBalanceVisible ? formatNumber(fiatAmount, { toFixed: 2 }) : lstrings.redacted_placeholder
  const handleToggleAccountBalanceVisibility = useHandler(() => {
    dispatch(toggleAccountBalanceVisibility())
  })

  const handleSend = useHandler(() => {
    Airship.show(bridge => <TransferModal depositOrSend="send" bridge={bridge} account={account} navigation={navigation} />).catch(() => {})
  })

  const handleDeposit = useHandler(() => {
    Airship.show(bridge => <TransferModal depositOrSend="deposit" bridge={bridge} account={account} navigation={navigation} />).catch(() => {})
  })

  const handleDigitLayout = useHandler((event: LayoutChangeEvent) => {
    setDigitHeight(event.nativeEvent.layout.height)
  })

  const balanceString = fiatSymbol.length !== 1 ? `${formattedFiat} ${fiatCurrencyCode}` : `${fiatSymbol} ${formattedFiat} ${fiatCurrencyCode}`
  const animateNumber = lt(fiatAmount, MAX_ANIMATED_AMOUNT)

  return (
    <EdgeCard>
      {/* For passing to the animated number. Do the measurement here to avoid flicker */}
      <Text style={[styles.balanceText, styles.measuredDigit]} onLayout={handleDigitLayout}>
        0
      </Text>
      <EdgeTouchableOpacity style={styles.balanceContainer} onPress={handleToggleAccountBalanceVisibility}>
        <View style={styles.titleContainer}>
          <EdgeText style={theme.cardTextShadow}>{lstrings.fragment_wallets_balance_text}</EdgeText>
          <IonIcon name={isBalanceVisible ? 'eye-off-outline' : 'eye-outline'} style={styles.eyeIcon} color={theme.iconTappable} size={theme.rem(1)} />
        </View>
        <View style={styles.balanceTextContainer}>
          {!exchangeRatesReady && zeroString(fiatAmount) ? (
            <EdgeText style={styles.balanceText}>{lstrings.loading}</EdgeText>
          ) : animateNumber ? (
            <AnimatedNumber digitHeight={digitHeight} numberString={balanceString} textStyle={styles.balanceText} />
          ) : (
            <EdgeText style={styles.balanceText}>{balanceString}</EdgeText>
          )}
        </View>
      </EdgeTouchableOpacity>
      {onViewAssetsPress == null ? null : (
        <EdgeTouchableOpacity style={styles.rightButtonContainer} onPress={onViewAssetsPress}>
          <EdgeText style={styles.tappableText}>{lstrings.view_assets}</EdgeText>
        </EdgeTouchableOpacity>
      )}

      <ButtonsView
        layout="row"
        secondary={{
          onPress: handleDeposit,
          label: lstrings.loan_fragment_deposit
        }}
        secondary2={{
          onPress: handleSend,
          label: lstrings.fragment_send_subtitle
        }}
      />
    </EdgeCard>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  balanceContainer: {
    margin: theme.rem(0.5)
  },
  titleContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: theme.rem(0.5)
  },

  rightButtonContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    position: 'absolute',
    top: 0,
    right: 0
  },
  tappableText: {
    fontSize: theme.rem(0.75),
    color: theme.iconTappable,
    margin: theme.rem(1),
    marginTop: theme.rem(1) + 3, // Fudge factor to align with the larger text on the left
    ...theme.cardTextShadow
  },

  eyeIcon: {
    alignSelf: 'center',
    marginLeft: theme.rem(0.25),
    marginRight: theme.rem(0),
    ...theme.cardTextShadow
  },
  balanceText: {
    fontSize: theme.rem(1.75),
    fontFamily: theme.fontFaceMedium,
    color: theme.primaryText,
    includeFontPadding: false,
    ...theme.cardTextShadow
  },
  balanceTextContainer: {
    height: theme.rem(2.5),
    flexDirection: 'row',
    alignItems: 'center'
  },
  showBalance: {
    fontSize: theme.rem(1.5),
    fontFamily: theme.fontFaceMedium
  },
  measuredDigit: { position: 'absolute', top: -999999 }
}))
