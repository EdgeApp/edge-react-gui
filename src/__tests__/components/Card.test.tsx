import { describe, expect, it, jest } from '@jest/globals'
import { Mock } from 'jest-mock' // Import Mock type from jest-mock if needed
import * as React from 'react'
import { View } from 'react-native'
import TestRenderer from 'react-test-renderer'

import { EdgeCard } from '../../components/cards/EdgeCard'
import { EdgeTouchableOpacity } from '../../components/common/EdgeTouchableOpacity'
import { EdgeText } from '../../components/themed/EdgeText'
import { FakeProviders } from '../../util/fake/FakeProviders'

const testColors = ['#4c669f', '#3b5998', '#192f6a']
const testIconUri = 'https://content.edge.app/currencyIconsV3/bitcoin/bitcoin.png'

describe('Card', () => {
  it('should render with gradient background', () => {
    const gradientProps = {
      colors: testColors,
      start: { x: 0, y: 0 },
      end: { x: 1, y: 1 }
    }
    const renderer = TestRenderer.create(
      <FakeProviders>
        <EdgeCard gradientBackground={gradientProps}>
          <EdgeText>Gradient Background</EdgeText>
        </EdgeCard>
      </FakeProviders>
    )
    expect(renderer.toJSON()).toMatchSnapshot()
  })

  it('should render with node background', () => {
    const nodeBackground = <View style={{ height: 100, width: 100, backgroundColor: testColors[0] }} />
    const renderer = TestRenderer.create(
      <FakeProviders>
        <EdgeCard nodeBackground={nodeBackground}>
          <EdgeText>Node Background</EdgeText>
        </EdgeCard>
      </FakeProviders>
    )
    expect(renderer.toJSON()).toMatchSnapshot()
  })

  it('should render with icon URI', () => {
    const renderer = TestRenderer.create(
      <FakeProviders>
        <EdgeCard icon={testIconUri}>
          <EdgeText>Icon</EdgeText>
        </EdgeCard>
      </FakeProviders>
    )
    expect(renderer.toJSON()).toMatchSnapshot()
  })

  it('should render sections correctly with multiple children', () => {
    const renderer = TestRenderer.create(
      <FakeProviders>
        <EdgeCard sections>
          <EdgeText>Section 1</EdgeText>
          <EdgeText>Section 2</EdgeText>
        </EdgeCard>
      </FakeProviders>
    )
    expect(renderer.toJSON()).toMatchSnapshot()
  })

  it('should handle press and long press events', () => {
    const mockOnPress: Mock<() => void | Promise<void>> = jest.fn()
    const mockOnLongPress: Mock<() => void | Promise<void>> = jest.fn()

    const testRenderer = TestRenderer.create(
      <FakeProviders>
        <EdgeCard onPress={mockOnPress} onLongPress={mockOnLongPress}>
          <EdgeText>Press Me</EdgeText>
        </EdgeCard>
      </FakeProviders>
    )

    testRenderer.root.findByType(EdgeTouchableOpacity).props.onPress()
    expect(mockOnPress).toHaveBeenCalled()

    testRenderer.root.findByType(EdgeTouchableOpacity).props.onLongPress()
    expect(mockOnLongPress).toHaveBeenCalled()

    expect(testRenderer.toJSON()).toMatchSnapshot()
  })

  it('should expand to fill its parent container', () => {
    const testRenderer = TestRenderer.create(
      <FakeProviders>
        <View style={{ flex: 1 }}>
          <EdgeCard>
            <EdgeText>Child Content</EdgeText>
          </EdgeCard>
        </View>
      </FakeProviders>
    )
    expect(testRenderer.toJSON()).toMatchSnapshot()
  })

  it('should allow child elements to expand within the card', () => {
    const testRenderer = TestRenderer.create(
      <FakeProviders>
        <EdgeCard>
          <View style={{ flex: 1, backgroundColor: testColors[0] }}>
            <EdgeText>Expanding Child</EdgeText>
          </View>
        </EdgeCard>
      </FakeProviders>
    )
    expect(testRenderer.toJSON()).toMatchSnapshot()
  })

  it('should expand both the card and its children to fill available space', () => {
    const testRenderer = TestRenderer.create(
      <FakeProviders>
        <View style={{ flex: 1 }}>
          <EdgeCard fill>
            <View style={{ flex: 1, backgroundColor: testColors[0] }}>
              <EdgeText>Expanding Child in Expanding Card</EdgeText>
            </View>
          </EdgeCard>
        </View>
      </FakeProviders>
    )
    expect(testRenderer.toJSON()).toMatchSnapshot()
  })
})
