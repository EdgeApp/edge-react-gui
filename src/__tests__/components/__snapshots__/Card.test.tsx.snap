// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render with loading props 1`] = `
<View
  style={
    Object {
      "width": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "borderColor": "rgba(255, 255, 255, .1)",
          "borderRadius": 4,
          "borderWidth": 1,
        },
        Object {
          "borderColor": "#F1AA19",
        },
        Object {
          "marginBottom": 247,
          "marginLeft": 247,
          "marginRight": 247,
          "marginTop": 247,
        },
        Object {
          "paddingBottom": 247,
          "paddingLeft": 247,
          "paddingRight": 247,
          "paddingTop": 247,
        },
      ]
    }
  >
    string
  </View>
</View>
`;
